`property` attribute with `@Consume` does not work

As per the use case outlined here:

https://camel.apache.org/manual/latest/pojo-consuming.html#_advanced_use_with_property_naming_convention

If you try to use `property`, then it results in an NPE:

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.2.0.CR1:build (quarkus-build) on project camel-quarkus-integration-test-messaging-common: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step org.apache.camel.quarkus.core.deployment.ConsumeProcessor#generateConsumeRoutes threw an exception: java.lang.NullPointerException
[ERROR]         at org.apache.camel.quarkus.core.deployment.ConsumeProcessor.generateConsumeRoutes(ConsumeProcessor.java:160)
```

Not sure whether its feasible to support such a feature at build time, so we could document it as unsupported.

@ppalaga WDYT?

--- a/extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/ConsumeProcessor.java
+++ b/extensions-core/core/deployment/src/main/java/org/apache/camel/quarkus/core/deployment/ConsumeProcessor.java
@@ -16,12 +16,16 @@
  */
 package org.apache.camel.quarkus.core.deployment;
 
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.List;
+import java.util.Optional;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import javax.decorator.Decorator;
 import javax.enterprise.context.ApplicationScoped;
@@ -50,6 +54,7 @@ import org.apache.camel.model.RoutesDefinition;
 import org.apache.camel.quarkus.core.ConsumeRecorder;
 import org.apache.camel.quarkus.core.deployment.spi.CamelContextBuildItem;
 import org.apache.camel.quarkus.support.common.CamelCapabilities;
+import org.apache.camel.util.StringHelper;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
 import org.jboss.jandex.AnnotationTarget.Kind;
@@ -157,13 +162,36 @@ public class ConsumeProcessor {
                 switch (target.kind()) {
                 case METHOD: {
                     final MethodInfo methodInfo = target.asMethod();
-                    final String uri = annot.value().asString();
+                    String uri = null;
+                    RuntimeValue<Object> runtimeUriOrEndpoint = null;
                     final ClassInfo declaringClass = methodInfo.declaringClass();
-                    if (uri.isEmpty()) {
-                        throw new IllegalStateException("@" + Consume.class.getName()
-                                + " requires a Camel endpoint URI in its value, e.g. @Consume(\"direct:myDirect\"): "
-                                + methodInfo
-                                + " in " + declaringClass.name());
+                    String beanName = namedValue(declaringClass);
+                    if (beanName == null) {
+                        beanName = ConsumeProcessor.uniqueBeanName(declaringClass);
+                    }
+                    if (annot.value() != null) {
+                        uri = annot.value().asString();
+                    } else if (annot.value("uri") != null) {
+                        uri = annot.value("uri").asString();
+                    } else if (annot.value("property") != null) {
+                        runtimeUriOrEndpoint = recorder.getEndpointUri(
+                                camelContext.getCamelContext(),
+                                beanName,
+                                findEndpointMethodName(annot.value("property").asString(), declaringClass, new ArrayList<>()));
+                    } else {
+                        runtimeUriOrEndpoint = recorder.getEndpointUri(
+                                camelContext.getCamelContext(),
+                                beanName,
+                                findEndpointMethodName(methodInfo.name(), declaringClass, new ArrayList<>()));
+                    }
+                    if (uri == null && runtimeUriOrEndpoint == null) {
+                        throw new IllegalStateException("@" + Consume.class.getName() + " on " + methodInfo + " in "
+                                + declaringClass.name()
+                                + " requires to specify a Camel Endpoint or endpoint URI through one of the following options:\n"
+                                + " * via @Consume value, e.g. @Consume(\"direct:myDirect\")\n"
+                                + " * via explicit @Consume property parameter, e.g. @Consume(property = \"myUriProperty\") and getMyUriProperty() or getMyUriPropertyEndpoint() in the same class must exist and return a String or org.apache.camel.Endpoint\n"
+                                + " * via naming convention: if your @Consume method is called myEvent or onMyEvent then getMyEvent() or getMyEventEndpoint() must exist in the same class and return a String or org.apache.camel.Endpoint\n"
+                                + "See https://camel.apache.org/manual/latest/pojo-consuming.html for more details");
                     }
                     if (!beanCapabilityAvailable) {
                         throw new IllegalStateException(
@@ -172,11 +200,8 @@ public class ConsumeProcessor {
                                         + methodInfo
                                         + " in " + declaringClass.name());
                     }
-                    String beanName = namedValue(declaringClass);
-                    if (beanName == null) {
-                        beanName = ConsumeProcessor.uniqueBeanName(declaringClass);
-                    }
-                    recorder.addConsumeRoute(routesDefinition, uri, beanName, methodInfo.name());
+                    recorder.addConsumeRoute(camelContext.getCamelContext(), routesDefinition, uri, runtimeUriOrEndpoint,
+                            beanName, methodInfo.name());
                     break;
                 }
                 default:
@@ -185,6 +210,28 @@ public class ConsumeProcessor {
             }
             recorder.addConsumeRoutesToContext(camelContext.getCamelContext(), routesDefinition);
         }
+    }
+
+    private String findEndpointMethodName(final String propertyName, ClassInfo declaringClass, List<String> triedMethods) {
+        /* Here we attempt to mimic what Camel does
+         * in org.apache.camel.impl.engine.CamelPostProcessorHelper.doGetEndpointInjection(Object, String, String) */
+        final String isGetter = "is" + StringHelper.capitalize(propertyName, false);
+        final String getGetter = "get" + StringHelper.capitalize(propertyName, false);
+        Optional<MethodInfo> method = Stream.of(isGetter, getGetter, isGetter + "Endpoint", getGetter + "Endpoint")
+                .peek(triedMethods::add)
+                .map(declaringClass::method)
+                .filter(m -> m != null)
+                .findFirst();
+        if (method.isEmpty()) {
+            if (propertyName.startsWith("on")) {
+                // retry but without the on as prefix
+                return findEndpointMethodName(propertyName.substring(2), declaringClass, triedMethods);
+            }
+            throw new IllegalStateException("Could not find any endpoint URI or Endpoint returning getter. Looked for " +
+                    triedMethods.stream().map(m -> m + "()").collect(Collectors.joining(", ")) + " in "
+                    + declaringClass.name().toString());
+        }
+        return method.get().name();
     }
 
     @BuildStep

--- a/extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/ConsumeRecorder.java
+++ b/extensions-core/core/runtime/src/main/java/org/apache/camel/quarkus/core/ConsumeRecorder.java
@@ -16,10 +16,17 @@
  */
 package org.apache.camel.quarkus.core;
 
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.stream.Stream;
+
 import io.quarkus.runtime.RuntimeValue;
 import io.quarkus.runtime.annotations.Recorder;
 import org.apache.camel.CamelContext;
 import org.apache.camel.Consume;
+import org.apache.camel.Endpoint;
+import org.apache.camel.NoTypeConversionAvailableException;
+import org.apache.camel.TypeConversionException;
 import org.apache.camel.model.Model;
 import org.apache.camel.model.RoutesDefinition;
 
@@ -34,9 +41,30 @@ public class ConsumeRecorder {
         return new RuntimeValue<RoutesDefinition>(routesDefinition);
     }
 
-    public void addConsumeRoute(RuntimeValue<RoutesDefinition> routesDefinition, String uri, String beanName, String method) {
+    public void addConsumeRoute(
+            RuntimeValue<CamelContext> camelContext,
+            RuntimeValue<RoutesDefinition> routesDefinition,
+            String uri,
+            RuntimeValue<Object> runtimeUriOrEndpoint,
+            String beanName,
+            String method) {
         final RoutesDefinition routes = routesDefinition.getValue();
-        routes.from(uri).bean(beanName, method);
+        if (uri != null) {
+            routes.from(uri).bean(beanName, method);
+        } else {
+            Object uriOrEndpoint = runtimeUriOrEndpoint.getValue();
+            if (uriOrEndpoint instanceof Endpoint) {
+                routes.from((Endpoint) uriOrEndpoint).bean(beanName, method);
+            } else {
+                try {
+                    final String uriOrRef = camelContext.getValue().getTypeConverter().mandatoryConvertTo(String.class,
+                            uriOrEndpoint);
+                    routes.from(uriOrRef).bean(beanName, method);
+                } catch (TypeConversionException | NoTypeConversionAvailableException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }
     }
 
     public void addConsumeRoutesToContext(RuntimeValue<CamelContext> camelContext,
@@ -47,6 +75,29 @@ public class ConsumeRecorder {
             camelContext.getValue().getExtension(Model.class).addRouteDefinitions(routes.getRoutes());
         } catch (Exception e) {
             throw new RuntimeException("Could not add routes to context", e);
+        }
+    }
+
+    public RuntimeValue<Object> getEndpointUri(RuntimeValue<CamelContext> camelContext, String beanName,
+            String endpointMethodName) {
+        /* Possible improvement: Instead of using reflection, we could generate this method at build time
+         * to call the bean method directly */
+        Object bean = camelContext.getValue().getRegistry().lookupByName(beanName);
+        Method method = null;
+        try {
+            Class<?> cl = bean.getClass();
+            do {
+                method = Stream.of(cl.getDeclaredMethods())
+                        .filter(m -> m.getName().equals(endpointMethodName) && m.getParameterCount() == 0)
+                        .findFirst()
+                        .orElse(null);
+                cl = cl.getSuperclass();
+            } while (method == null && cl != Object.class);
+            Object result = method.invoke(bean);
+            return new RuntimeValue<>(result);
+        } catch (SecurityException | IllegalAccessException | IllegalArgumentException
+                | InvocationTargetException e) {
+            throw new RuntimeException(e);
         }
     }
 

--- a/integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/consume/ConsumeAnnotationWithExplicitPropertyBean.java
+++ b/integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/consume/ConsumeAnnotationWithExplicitPropertyBean.java
@@ -1,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.quarkus.component.bean.consume;
+
+import org.apache.camel.Consume;
+
+/**
+ * A bean having a method annotated with {@code @Consume} with an explicit {@link Consume#property()} value.
+ */
+public class ConsumeAnnotationWithExplicitPropertyBean {
+
+    @Consume(property = "myEndpoint")
+    public String consumeAnnotation(String name) {
+        return "Consumed " + name + " via direct:consumeAnnotationWithExplicitProperty";
+    }
+
+    public String getMyEndpoint() {
+        return "direct:consumeAnnotationWithExplicitProperty";
+    }
+
+}

--- a/integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/consume/ConsumeAnnotationWithImplicitEndpointGetterBean.java
+++ b/integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/consume/ConsumeAnnotationWithImplicitEndpointGetterBean.java
@@ -1,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.quarkus.component.bean.consume;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.apache.camel.Consume;
+import org.apache.camel.Endpoint;
+import org.apache.camel.EndpointInject;
+import org.apache.camel.component.direct.DirectEndpoint;
+
+/**
+ * A bean having a method annotated with {@code @Consume} with a fallback endpoint URI getter whose name ends with
+ * {@code Endpoint} and whose return type is {@link Endpoint}.
+ */
+@ApplicationScoped
+public class ConsumeAnnotationWithImplicitEndpointGetterBean {
+    @EndpointInject("direct:consumeAnnotationWithImplicitEndpointGetter")
+    DirectEndpoint directEndpoint;
+
+    @Consume
+    public String specialEvent(String name) {
+        return "Consumed " + name + " via direct:consumeAnnotationWithImplicitEndpointGetter";
+    }
+
+    public Endpoint getSpecialEventEndpoint() {
+        return directEndpoint;
+    }
+
+}

--- a/integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/consume/ConsumeAnnotationWithImplicitGetterBean.java
+++ b/integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/consume/ConsumeAnnotationWithImplicitGetterBean.java
@@ -1,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.quarkus.component.bean.consume;
+
+import org.apache.camel.Consume;
+
+/**
+ * A bean having a method annotated with {@code @Consume} with an implicit endpoint URI getter.
+ */
+public class ConsumeAnnotationWithImplicitGetterBean {
+
+    @Consume
+    public String specialEvent(String name) {
+        return "Consumed " + name + " via direct:consumeAnnotationWithImplicitGetter";
+    }
+
+    public String getSpecialEvent() {
+        return "direct:consumeAnnotationWithImplicitGetter";
+    }
+
+}

--- a/integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/consume/ConsumeAnnotationWithImplicitOnGetterBean.java
+++ b/integration-test-groups/foundation/bean/src/main/java/org/apache/camel/quarkus/component/bean/consume/ConsumeAnnotationWithImplicitOnGetterBean.java
@@ -1,0 +1,36 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.quarkus.component.bean.consume;
+
+import org.apache.camel.Consume;
+
+/**
+ * A bean having a method annotated with {@code @Consume} with a fallback endpoint URI getter whose name results from
+ * removing {@code on} prefix from the <code>@Consume</code> method.
+ */
+public class ConsumeAnnotationWithImplicitOnGetterBean {
+
+    @Consume
+    public String onSpecialEvent(String name) {
+        return "Consumed " + name + " via direct:consumeAnnotationWithImplicitOnGetter";
+    }
+
+    public String getSpecialEvent() {
+        return "direct:consumeAnnotationWithImplicitOnGetter";
+    }
+
+}

--- a/integration-test-groups/foundation/bean/src/test/java/org/apache/camel/quarkus/component/bean/ConsumeAnnotationTest.java
+++ b/integration-test-groups/foundation/bean/src/test/java/org/apache/camel/quarkus/component/bean/ConsumeAnnotationTest.java
@@ -36,6 +36,46 @@ public class ConsumeAnnotationTest {
     }
 
     @Test
+    public void consumeAnnotationWithExplicitProperty() {
+        RestAssured.given()
+                .contentType(ContentType.TEXT)
+                .body("bar")
+                .post("/bean/route/consumeAnnotationWithExplicitProperty")
+                .then()
+                .body(equalTo("Consumed bar via direct:consumeAnnotationWithExplicitProperty"));
+    }
+
+    @Test
+    public void consumeAnnotationWithImplicitGetter() {
+        RestAssured.given()
+                .contentType(ContentType.TEXT)
+                .body("baz")
+                .post("/bean/route/consumeAnnotationWithImplicitGetter")
+                .then()
+                .body(equalTo("Consumed baz via direct:consumeAnnotationWithImplicitGetter"));
+    }
+
+    @Test
+    public void consumeAnnotationWithImplicitEndpointGetter() {
+        RestAssured.given()
+                .contentType(ContentType.TEXT)
+                .body("boo")
+                .post("/bean/route/consumeAnnotationWithImplicitEndpointGetter")
+                .then()
+                .body(equalTo("Consumed boo via direct:consumeAnnotationWithImplicitEndpointGetter"));
+    }
+
+    @Test
+    public void consumeAnnotationWithImplicitOnGetter() {
+        RestAssured.given()
+                .contentType(ContentType.TEXT)
+                .body("boo")
+                .post("/bean/route/consumeAnnotationWithImplicitOnGetter")
+                .then()
+                .body(equalTo("Consumed boo via direct:consumeAnnotationWithImplicitOnGetter"));
+    }
+
+    @Test
     public void defaultNamedConsumeAnnotation() {
         RestAssured.given()
                 .contentType(ContentType.TEXT)

